first iteration of migration docs for vue3 (#12380)
rancher/dashboard: docusaurus/docs/extensions/rancher-2.10-support.md

# Rancher 2.10.0 update

## Breaking Changes in 2.10.0

> With the update Rancher UI, Shell package and the UI Extensions from Vue2 to Vue3, this has introduced breaking changes to the UI Extensions. requiring that you must update your existing extensions in order to be compliant with the Rancher `2.10` UI framework.

Reviewer: ```suggestion
Following the updates to the Rancher UI, Shell package, and the UI Extensions from Vue2 to Vue3, we have introduced breaking changes to the UI Extensions. These changes require you to update your existing extensions in order to be compliant with the Rancher `2.10` UI framework.
```

## How to proceed with your extension update

We have developed a migration script that will allow extension developers to easily migrate their extensions to the minimum required setup in order to be compliant with this breaking change.

Notably we update the Shell package to version `3.x.x`, which is now a requirement to be compliant with Rancher `2.10`, but we also update some of the core files of you app skeleton in order to be compatible with the new node version now in use `v20`. To find out more about the support matrix for Shell versions in regards to Rancher versions, check the support matrix [here](./support-matrix#shell-support-matrix).

> On top of this, the migration script also does some of the basic changes in terms of syntax needed to work with Vue3, however we must refer that this feature is considered **experimental** and therefore we **don't** provide support for the syntax migration. We recommend that you become familiar with the changes between Vue2 and Vue3 by following the official documentation [here](https://v3-migration.vuejs.org/).

Reviewer: ```suggestion
On top of this, the migration script also does some of the basic changes in terms of syntax needed to work with Vue3. However, please note that this feature is **experimental** and therefore **unsupported** for syntax migration. We recommend familiarizing yourself with the differences between Vue2 and Vue3 by following the official documentation [here](https://v3-migration.vuejs.org/).
```

- Attribute Coercion Behavior: https://v3-migration.vuejs.org/breaking-changes/attribute-coercion.html#attribute-coercion-behavior
- Inline Template Attribute: https://v3-migration.vuejs.org/breaking-changes/inline-template-attribute.html#inline-template-attribute

## New mandatory annotation

One of the new changes that we've introduced is that for a published extension to be loaded into a Rancher `2.10` system, the published helm chart for a particular version **must** include the `catalog.cattle.io/ui-extensions-version` annotation, being the minimum version needed `>= 3.0.0`.

We also recommend that you update the `catalog.cattle.io/rancher-version` to `>= 2.10.0`.

You should check that your extension `pkg/<-YOUR EXTENSION->/package.json` shows something similar to:

```json
{
  "name": "your-extension",
  "description": "your-extension description",
  "version": "1.2.1",
  "rancher": {
    "annotations": {
      "catalog.cattle.io/rancher-version": ">= 2.10.0",
      "catalog.cattle.io/ui-extensions-version": ">= 3.0.0"
    }
  },
  ....
}
```

## How to maintain different extension versions

> In terms of development procedure, we do recommend that you branch your extension repository so that you can have one branch tracking the Rancher "2.9" and the Rancher 2.10 “worlds”, if your extension is affected by the changes. 

Reviewer: ```suggestion
In terms of development procedure, we recommend branching your extension repository to manage compatibility with both Rancher 2.9 and Rancher 2.10, if your extension is impacted by version-specific changes.
```

You can check the example of `Elemental` where branch `main` is tracking the **Rancher 2.10 compliant version** and `release-2.9.x` branch is tracking everything Rancher "2.9".

> The difference between the two branches is some annotations and of course the shell versions being used.

Reviewer: ```suggestion
The difference between the two branches are some annotations and of course the shell versions being used.
```
